### PR TITLE
Devise login to app

### PR DESCRIPTION
**Pull request by danielsteenberg**
Created devise and google login for domain

### CODE REVIEW

File: app/assets/javascripts/sessions.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (jamieandyou) [inline app/assets/javascripts/sessions.coffee]: Is this file necessary?

File: app/assets/stylesheets/homepage.scss [hunk 2]
```diff
@@ -15,3 +15,9 @@
 .row {
   float: right;
 }
+
+.page-header{
+	h1{
+		text-align: center;
+	}
+}
```
Comment (jamieandyou) [inline app/assets/stylesheets/homepage.scss:23]: tab spacing on this file too big

File: app/assets/stylesheets/inputform.scss [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
+    padding: 10px;
+    border: 1px solid #999;
+    height: 5%;
+    border: inset 1px solid #333;
+    -webkit-box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    -moz-box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    background: -webkit-linear-gradient(bottom, #CCCCCC, #EEEEEE 175px);
+    background: -moz-linear-gradient(bottom, #CCCCCC, #EEEEEE 175px);
+    background: linear-gradient(bottom, #CCCCCC, #EEEEEE 175px);
```
Comment (jamieandyou) [inline app/assets/stylesheets/inputform.scss]: lower case colour codes and only use 3 letters instead of 6

File: app/assets/stylesheets/inputform.scss [hunk 4]
```diff
@@ -1,18 +1,23 @@
 .form {
-  border-radius: 10px;
```
Comment (jamieandyou) [inline app/assets/stylesheets/inputform.scss:2]: Indentation on this whole file is now too big

File: app/assets/stylesheets/modern-business.scss [hunk 5]
```diff
@@ -5,6 +5,7 @@ body {
 
 body {
   padding-top: 50px;
+  background-color: e8e8e8;
```
Comment (jamieandyou) [inline app/assets/stylesheets/modern-business.scss]: Missing the # in front of the code

File: app/assets/stylesheets/sessions.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (jamieandyou) [inline app/assets/stylesheets/sessions.scss]: Do you need this file?

File: app/controllers/omniauth_callbacks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController   
+def google_oauth2
+
```
Comment (jamieandyou) [inline app/controllers/omniauth_callbacks_controller.rb]: can remove this empty line

File: app/models/user.rb [hunk 8]
```diff
@@ -3,4 +3,34 @@ class User < ActiveRecord::Base
...
+                   with: /\A[\w+-.]+@getsmarter.co.za\z/i
+                   }
+
+  def self.find_for_google_oauth2(access_token, signed_in_resource=nil)
+    data = access_token.info
+    user = User.where(:provider => access_token.provider, :uid => access_token.uid ).first
+    if user
+      return user
+    else
+      registered_user = User.where(:email => access_token.info.email).first
```
Comment (jamieandyou) [inline app/models/user.rb]: Change to the format without the =>

File: app/assets/javascripts/sessions.coffee (hunk 1 above)
Comment (danielsteenberg) [inline app/assets/javascripts/sessions.coffee]: nope been deleted now

File: app/models/user.rb [hunk 9]
```diff
@@ -3,4 +3,34 @@ class User < ActiveRecord::Base
...
+      if registered_user
+        return registered_user
+      else
+        user = User.create(email: data["email"],
+          provider:access_token.provider,
+          email: data["email"],
+          uid: access_token.uid ,
+          password: Devise.friendly_token[0,20],
```
Comment (jamieandyou) [inline app/models/user.rb]: Remove comma at the end of this line, unnecessary

File: app/assets/stylesheets/homepage.scss (hunk 2 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/homepage.scss:23]: fine on my side but set tab spaced to line spaces, so should solve the problem on next pushup

File: app/assets/stylesheets/inputform.scss (hunk 3 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/inputform.scss]: changed and sorted

File: app/views/devise/registrations/new.html.erb [hunk 10]
```diff
@@ -25,5 +27,6 @@
     <%= f.submit "Sign up" %>
   </div>
 <% end %>
+</div>
 
 <%= render "devise/shared/links" %>
```
Comment (jamieandyou) [inline app/views/devise/registrations/new.html.erb]: Why isn't this file slim?

(Same to all other .erb files)

File: app/views/sessions/create.html.slim [hunk 11]
```diff
@@ -0,0 +1,2 @@
+h1 Sessions#create
+p Find me in app/views/sessions/create.html.slim
```
Comment (jamieandyou) [inline app/views/sessions/create.html.slim]: What is the purpose of this page?

File: app/views/sessions/destroy.html.slim [hunk 12]
```diff
@@ -0,0 +1,2 @@
+h1 Sessions#destroy
+p Find me in app/views/sessions/destroy.html.slim
```
Comment (jamieandyou) [inline app/views/sessions/destroy.html.slim]: What is the purpose of this page?

File: app/assets/stylesheets/inputform.scss (hunk 4 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/inputform.scss:2]: fine on my side but did the convert tab to spaces and should solve this issue.

File: app/assets/stylesheets/modern-business.scss (hunk 5 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/modern-business.scss]: cool, thanks, added it now

File: spec/helpers/sessions_helper_spec.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+# describe SessionsHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       helper.concat_strings("this","that").should == "this that"
+#     end
+#   end
+# end
+describe SessionsHelper do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (jamieandyou) [inline spec/helpers/sessions_helper_spec.rb]: Remove these specs if they're not being used

File: app/assets/stylesheets/sessions.scss (hunk 6 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/sessions.scss]: deleted

File: app/controllers/omniauth_callbacks_controller.rb (hunk 7 above)
Comment (danielsteenberg) [inline app/controllers/omniauth_callbacks_controller.rb]: cool sorted

File: app/models/user.rb (hunk 8 above)
Comment (danielsteenberg) [inline app/models/user.rb]: sorted

File: app/models/user.rb (hunk 9 above)
Comment (danielsteenberg) [inline app/models/user.rb]: sorted

File: app/views/devise/registrations/new.html.erb (hunk 10 above)
Comment (danielsteenberg) [inline app/views/devise/registrations/new.html.erb]: will convert it over will take some time to do that tonight

File: app/models/user.rb [hunk 14]
```diff
@@ -3,4 +3,34 @@ class User < ActiveRecord::Base
...
+           presence: true,
+           uniqueness: true,
+           format: {
+                   message: 'You Do Not Belong To The GetSmarter Domain',
+                   with: /\A[\w+-.]+@getsmarter.co.za\z/i
+                   }
+
+  def self.find_for_google_oauth2(access_token, signed_in_resource=nil)
+    data = access_token.info
+    user = User.where(:provider => access_token.provider, :uid => access_token.uid ).first
```
Comment (jamieandyou) [inline app/models/user.rb]: change all the =>

File: app/views/sessions/create.html.slim (hunk 11 above)
Comment (danielsteenberg) [inline app/views/sessions/create.html.slim]: Deleted

File: app/views/sessions/destroy.html.slim (hunk 12 above)
Comment (danielsteenberg) [inline app/views/sessions/destroy.html.slim]: Deleted

File: spec/helpers/sessions_helper_spec.rb (hunk 13 above)
Comment (danielsteenberg) [inline spec/helpers/sessions_helper_spec.rb]: deleted pages controllers and models for sessions not being used

File: app/views/devise/registrations/new.html.erb (hunk 10 above)
Comment (danielsteenberg) [inline app/views/devise/registrations/new.html.erb]: Done

File: app/views/devise/mailer/reset_password_instructions.html.slim [hunk 15]
```diff
@@ -0,0 +1,9 @@
+p Hello 
+	= @resource.email
```
Comment (jamieandyou) [inline app/views/devise/mailer/reset_password_instructions.html.slim]: Indentation too much

File: app/views/devise/mailer/reset_password_instructions.html.slim [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+
+p Someone has requested a link to change your password. You can do this through the link below:
+
+p = link_to 'Change my password', edit_password_url(@resource, reset_password_token: @token)
```
Comment (jamieandyou) [inline app/views/devise/mailer/reset_password_instructions.html.slim]: Put from the = down a line for consistency with your other files

File: app/views/devise/passwords/edit.html.slim [hunk 17]
```diff
@@ -0,0 +1,24 @@
+h2 Change your password
+
+= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f|
+= devise_error_messages!
+= f.hidden_field :reset_password_token
```
Comment (jamieandyou) [inline app/views/devise/passwords/edit.html.slim]: These two lines need to be indented in the form

File: app/views/devise/passwords/edit.html.slim [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+
+  .field
+    = f.label :password, "New password" 
+    br
+      = if @minimum_password_length
```
Comment (jamieandyou) [inline app/views/devise/passwords/edit.html.slim]: This should be a hyphen - as far as I know. Because you aren't outputting the code onto the page

File: app/views/devise/registrations/edit.html.slim [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f|
+  = devise_error_messages!
+
+  .field
+    = f.label :email 
+    br
+    = f.email_field :email, autofocus: true
+
+  if devise_mapping.confirmable? && resource.pending_reconfirmation?
```
Comment (jamieandyou) [inline app/views/devise/registrations/edit.html.slim]: Needs a hyphen in front of this

File: app/views/devise/registrations/new.html.slim [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+    = devise_error_messages!
+
+    .field
+      = f.label :email 
+      br
+      = f.email_field :email, autofocus: true
+
+    .field
+      = f.label :password 
+      -if @minimum_password_length
```
Comment (jamieandyou) [inline app/views/devise/registrations/new.html.slim]: What is this if doing? The code beneath an if must always be indented in

File: app/views/devise/sessions/new.html.slim [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    .field
+      = f.label :email
+      = f.email_field :email, autofocus: true
+
+    .field
+      = f.label :password
+      = f.password_field :password, autocomplete: "off"
+
+    -if devise_mapping.rememberable?
+    .field
```
Comment (jamieandyou) [inline app/views/devise/sessions/new.html.slim]: Indent

File: spec/helpers/sessions_helper_spec.rb (hunk 13 above)
Comment (jamieandyou) [inline spec/helpers/sessions_helper_spec.rb]: These are all still in this PR. I thought you said you deleted them

File: app/views/devise/mailer/reset_password_instructions.html.slim (hunk 15 above)
Comment (danielsteenberg) [inline app/views/devise/mailer/reset_password_instructions.html.slim]: sorted

File: app/views/devise/mailer/reset_password_instructions.html.slim (hunk 16 above)
Comment (danielsteenberg) [inline app/views/devise/mailer/reset_password_instructions.html.slim]: sorted

File: app/views/devise/passwords/edit.html.slim (hunk 17 above)
Comment (danielsteenberg) [inline app/views/devise/passwords/edit.html.slim]: sorted

File: app/views/devise/passwords/edit.html.slim (hunk 18 above)
Comment (danielsteenberg) [inline app/views/devise/passwords/edit.html.slim]: sorted

File: app/views/devise/registrations/new.html.slim (hunk 20 above)
Comment (danielsteenberg) [inline app/views/devise/registrations/new.html.slim]: lines underneath is indented

File: app/views/devise/sessions/new.html.slim (hunk 21 above)
Comment (danielsteenberg) [inline app/views/devise/sessions/new.html.slim]: sorted

File: app/assets/stylesheets/inputform.scss [hunk 22]
```diff
@@ -1,18 +1,23 @@
...
+    -webkit-box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    -moz-box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    box-shadow: 0px 0px 8px rgba(0, 0, 0, 0.3);
+    background: -webkit-linear-gradient(bottom, #ccc, #eee 175px);
+    background: -moz-linear-gradient(bottom, #ccc, #eee 175px);
+    background: linear-gradient(bottom, #ccc, #eee 175px);
+    margin: auto;
+    position: relative;
+    width: 40%;
+    height: 55%;
```
Comment (jamieandyou) [inline app/assets/stylesheets/inputform.scss:15]: sorry about this man, but I've just seen you've put height twice in this

File: app/views/devise/passwords/edit.html.slim [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+
+= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f|
+  = devise_error_messages!
+  = f.hidden_field :reset_password_token
+
+  .field
+    = f.label :password, "New password" 
+    br
+      - if @minimum_password_length
+    em
```
Comment (jamieandyou) [inline app/views/devise/passwords/edit.html.slim]: This should be tabbed in under the if

File: spec/views/sessions/destroy.html.slim_spec.rb [hunk 24]
```diff
@@ -0,0 +1,5 @@
+require 'spec_helper'
+
+describe "sessions/destroy.html.slim" do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (jamieandyou) [inline spec/views/sessions/destroy.html.slim_spec.rb]: You still haven't deleted these unused specs.

File: app/assets/stylesheets/inputform.scss (hunk 22 above)
Comment (danielsteenberg) [inline app/assets/stylesheets/inputform.scss:15]: cool deleted the one.

File: spec/views/sessions/destroy.html.slim_spec.rb (hunk 24 above)
Comment (danielsteenberg) [inline spec/views/sessions/destroy.html.slim_spec.rb]: deleted

File: app/views/devise/passwords/edit.html.slim (hunk 23 above)
Comment (danielsteenberg) [inline app/views/devise/passwords/edit.html.slim]: sorted

File: app/views/devise/passwords/edit.html.slim [hunk 25]
```diff
@@ -8,7 +8,7 @@ h2 Change your password
     = f.label :password, "New password" 
     br
       - if @minimum_password_length
-    em 
+      em 
       = @minimum_password_length  (characters minimum)
```
Comment (jamieandyou) [inline app/views/devise/passwords/edit.html.slim:12]: This line needs to be tabbed in as well Dan, it seems like you're not actually seeing how this stuff renders on the pages.
And it also shows that you haven't really checked out how slim works with it's indentation.

File: spec/models/google_user_spec.rb [hunk 26]
```diff
@@ -0,0 +1,5 @@
+require 'spec_helper'
+
+describe GoogleUser do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
Comment (jamieandyou) [inline spec/models/google_user_spec.rb]: You haven't deleted this spec either

File: spec/helpers/sessions_helper_spec.rb (hunk 13 above)
Comment (jamieandyou) [inline spec/helpers/sessions_helper_spec.rb]: You still haven't deleted this spec

File: app/views/devise/registrations/edit.html.slim [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f|
+  = devise_error_messages!
+
+  .field
+    = f.label :email 
+    br
+    = f.email_field :email, autofocus: true
+
+  -if devise_mapping.confirmable? && resource.pending_reconfirmation?
```
Comment (jamieandyou) [inline app/views/devise/registrations/edit.html.slim]: Also, find all your -if and make them - if with a space

**Comment by jamieandyou** [comment]
👍
